pkg/logs/sender/sender.go

```diff
@@ -6,220 +6,125 @@
-	lastError    error
-	trackErrors  bool
+	bufferSize   int
 }
 
-// NewSingleSender returns a new sender.
-func NewSingleSender(inputChan chan *message.Message, outputChan chan *message.Message, destinations *client.Destinations, strategy Strategy) *SingleSender {
-	return &SingleSender{
+// NewSender returns a new sender.
+func NewSender(inputChan chan *message.Payload, outputChan chan *message.Payload, destinations *client.Destinations, bufferSize int) *Sender {
+	return &Sender{
 		inputChan:    inputChan,
 		outputChan:   outputChan,
-		hasError:     make(chan bool, 1),
 		destinations: destinations,
-		strategy:     strategy,
 		done:         make(chan struct{}),
-		trackErrors:  false,
+		bufferSize:   bufferSize,
 	}
 }
 
 // Start starts the sender.
-func (s *SingleSender) Start() {
+func (s *Sender) Start() {
 	go s.run()
 }
 
 // Stop stops the sender,
 // this call blocks until inputChan is flushed
-func (s *SingleSender) Stop() {
+func (s *Sender) Stop() {
 	close(s.inputChan)
 	<-s.done
 }
 
-// Flush sends synchronously the messages that this sender has to send.
-func (s *SingleSender) Flush(ctx context.Context) {
-	s.strategy.Flush(ctx)
-}
+func (s *Sender) run() {
+	reliableDestinations := buildDestinationStates(s.destinations.Reliable, s.outputChan, s.bufferSize)
 
-func (s *SingleSender) run() {
-	defer func() {
-		s.done <- struct{}{}
-	}()
-	s.strategy.Send(s.inputChan, s.outputChan, s.send)
-}
+	sink := additionalDestinationsSink(s.bufferSize)
+	additionalDestinations := buildDestinationStates(s.destinations.Additionals, sink, s.bufferSize)
+
+	for payload := range s.inputChan {
 
-// send sends a payload to multiple destinations,
-// it will forever retry for the main destination unless the error is not retryable
-// and only try once for additional destinations.
-func (s *SingleSender) send(payload []byte) error {
-	for {
-		err := s.destinations.Main.Send(payload)
-		if err != nil {
-			if s.trackErrors && s.lastError == nil {
-				s.hasError <- true
+		sent := false
+		for !sent {
+			for _, destState := range reliableDestinations {
+				if !destState.destination.GetIsRetrying() {
+					destState.input <- payload
+					sent = true
+				}
 			}
-			s.lastError = err
 
-			metrics.DestinationErrors.Add(1)
-			metrics.TlmDestinationErrors.Inc()
-			if _, ok := err.(*client.RetryableError); ok {
+			if !sent {
+				// Throttle the poll loop while waiting for a status change from one
+				// of the destinations. This will only happen when all reliable destinations
+				// are blocked so logs have no where to go.
+				time.Sleep(100 * time.Millisecond)
+			}
+		}
 
-				// could not send the payload because of a client issue,
-				// let's retry
-				continue
+		for _, destState := range reliableDestinations {
+			// if an endpoint is stuck in the previous step, try to buffer the payloads if we have room to mitigate
+			// loss on intermittent failures.
+			if destState.destination.GetIsRetrying() {
```

Comment:
I think there's an implicit assumption here that `GetIsRetrying` here is equal to `GetIsRetrying` on line 64, so that this case triggers only in the situation where at least one reliable destination was able to send but this one was not.  That's not quite accurate -- it may not have been retrying earlier, and be retrying now, in which case the message may be double-queued.

Reply:
Good point. In a previous version I was caching the retry value in `destinationState` so it dones't get polled twice. I'll add that back.

Reply:
It seems like GetIsRetrying is an approximation to "channel is full" here (or more accurately, `GetIsRetrying() ~= len(chan) > 0`?)

Would it be easier to do a non-blocking push to all of them, and just record which succeeded and which fell through due to a full channel?  Then if all fell through, wait and retry, otherwise try sending to the additionals.

Reply:
> It seems like GetIsRetrying is an approximation to "channel is full"

It does a little more than that. `GetIsRetrying` can return true even when the channel is not full. it just means that one of the transactions has failed with a retry-able error and stuck in a retry loop. in that case we don't want to consider it blocking anymore and do a best effort write to channel to buffer messages. This helps mitigate loss of logs when there are intermittent failures.

Reply:
Refactored this - explained in another comment.